rhine-terminal/src/FRP/Rhine/Terminal.hs

```diff
@@ -0,0 +1,165 @@
+{- | Wrapper to write @terminal@ applications in Rhine, using concurrency.
+-}
+
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TypeFamilies #-}
+module FRP.Rhine.Terminal
+  ( TerminalConcT
+  , TerminalEventClock (..)
+  , launchTerminal
+  , flowTerminal
+  , terminalConcurrently
+  )
+  where
+
+-- base
+import Prelude hiding (putChar)
+import Control.Concurrent ()
+import Control.Concurrent.MVar ()
+import Data.IORef ()
+
+-- time
+import Data.Time.Clock ( getCurrentTime )
+
+-- transformers
+import Control.Monad.Trans.Class ( MonadTrans )
+import Control.Monad.Trans.Reader ( ReaderT(..), ask )
+
+-- terminal
+-- terminal
+import System.Terminal
+import System.Terminal.Internal ( LocalTerminal )
+
+-- rhine
+import FRP.Rhine
+    ( MonadIO(..),
+      constM,
+      liftTransS,
+      flow,
+      concurrently,
+      Arrow(first),
+      Clock(..),
+      HoistClock(..),
+      GetClockProxy,
+      In,
+      Out,
+      Schedule(..),
+      Rhine,
+      UTCTime )
+import Control.Monad.Catch (MonadMask, MonadCatch, MonadThrow)
+
+-- | Wraps the LocalTerminal value needed for communication with the @terminal@ backend.
+newtype TerminalConcT m a = TerminalConcT
+  { unTerminalConcT :: ReaderT LocalTerminal m a }
+  deriving (Functor, Applicative, Monad, MonadTrans, MonadIO, MonadThrow, MonadCatch, MonadMask)
+
+instance MonadInput (TerminalConcT IO) where
+  awaitWith = awaitWith
+
+instance MonadPrinter (TerminalConcT IO) where
+  putLn = putLn
+  putChar = putChar
+  putText = putText
+  getLineWidth = getLineWidth
+
+instance MonadScreen (TerminalConcT IO) where
+  getWindowSize = getWindowSize
+  moveCursorUp = moveCursorUp
+  moveCursorDown = moveCursorDown
+  moveCursorForward = moveCursorForward
+  moveCursorBackward = moveCursorBackward
+  getCursorPosition = getCursorPosition
+  setCursorPosition = setCursorPosition
+  setCursorRow = setCursorRow
+  setCursorColumn = setCursorColumn
+  saveCursor = saveCursor
+  restoreCursor = restoreCursor
+  insertChars = insertChars
+  deleteChars = deleteChars
+  eraseChars = eraseChars
+  insertLines = insertLines
+  deleteLines = deleteLines
+  eraseInLine = eraseInLine
+  eraseInDisplay = eraseInDisplay
+  showCursor = showCursor
+  hideCursor = hideCursor
+  setAutoWrap = setAutoWrap
+  setAlternateScreenBuffer = setAlternateScreenBuffer
+
+-- * Terminal clock in 'IO'
+
+-- | Concurrently block on @gloss@ events.
+data TerminalEventClock = TerminalEventClock
+
+instance (MonadIO m, MonadMask m) => Clock (TerminalConcT m) TerminalEventClock where
```

Comment:
Can you leave the `TerminalConcT` out of the clock definition? 
```suggestion
instance (MonadIO m, MonadMask m, MonadReader Terminal m, MonadInput m) => Clock m TerminalEventClock where
```

This has the advantage that we don't need to pin down the transformer stack when using the clock.

Reply:
Yes, that makes a lot of sense.

Reply:
I am running into some trouble further down though when trying to write `terminalConcurrently` in this style because concurrently pins `IO`. 
```haskell
concurrently
  :: ( Clock IO cl1, Clock IO cl2
     , Time cl1 ~ Time cl2
     )
  => Schedule IO cl1 cl2
```

Should there be (can there be) a concurrently which is over `MonadIO m`?

Reply:
Ok sounds good. I had the feeling it wasn't worth the extra dependency.